warehouse/macaroons/models.py

```diff
@@ -71,6 +76,38 @@ class Macaroon(db.Model):
         JSONB, server_default=sql.text("'{}'")
     )
 
+    # The caveats that were attached to this Macaroon when we generated it.
+    #
+    # NOTE: The cannonical, authoritative list of caveats that a Macaroon has is
+    #       contained within the Macaroon itself. These caveats are *not* used
+    #       during the authorization or authentication process, and are retained
+    #       merely for informational reasons.
+    #
+    # NOTE: Users can add additional caveats at any time without communicating
+    #       those additional caveats to us, so those would not be reflected in
+    #       this data.
+    #
+    # NOTE: Originally we only stored the "permissions" caveat, which maps to
+    #       the RequestUser and ProjectName caveat, while that data was mapped
+    #       into these caveats, there may have been additional caveats (such as
+    #       expiration) that were attached to the Macaroon during generation
+    #       that is missing from this column for older Macaroons.
+    #
+    _caveats: Mapped[list | None] = mapped_column("caveats", JSONB)
```

Comment:
There's something bothering me about this data type, and I'm not sure why, so let me write out some thoughts.

While a `json(b)`-type columns can absolutely contain a list instead of a dict, to date we've used only dict (json object). This sets the column to be a list (json array), and in my head that would be a postgres `jsonb[]` data type instead - as that would constrain the type to only ever allow arrays of json objects at the DB level. 

I think the right way to express that is:

```python
from sqlalchemy.dialects.postgresql import ARRAY

_caveats: Mapped[list | None] = mapped_column("caveats", ARRAY(JSONB))
```
and the resulting migration gives us a db-level `jsonb[]` type.

I don't know if there's DB-space efficiency or index/query time efficiency, but I think having a column type constrained to what is expected to be there is better.

There's an imperfect example showing some differences here: https://stackoverflow.com/a/56625335/244037 (reference is from 2019, things may have changed a bit, so worth revalidating).

But if that materially changes the query-time ergonomics, or makes things even more confusing, then I'm happy to pass on this.

Reply:
I considered doing `jsonb[]` but ended up not doing that because it means that to query the column you need to mix PostgreSQL array functions with jsonb functions, and I thought that the combination of those two would be more confusing than just using a `jsonb` as an array.

I'm happy to swap this to a `jsonb[]` though if people prefer, I don't really have a preference one way or another, I was on the fence and just sorta went with what felt right.

Reply:
I've gone ahead and done this.

Reply:
Actually, I'm going to undo it because it makes SQLAlchemy barf when you have multiple caveats.

The problem is that SQLAlchemy ends up generating SQL like this:

```
 "{{4,"\"7804c7db-c87d-4fee-a254-e0baafd06c59\"","null"},{2,"[\"2767a5b8-85f6-4bf4-9adc-6ed507df6c57\"]"},{0,900,0}}"
 ```

And PostgreSQL sees that as a multi dimensional array, not as an array of JSONB.

Reply:
Okay, that's icky. I wonder if there's a better way to express the concept in SQLAlchemy to have the queries make more sense? In any case, no need to pursue this further - thanks for trying it out!